src/extending-phpunit.rst

```diff
@@ -6,100 +6,175 @@
 
-        <extensions>
-            <bootstrap class="Vendor\ExampleExtensionForPhpunit\Extension">
-                <parameter name="message" value="the-message"/>
-            </bootstrap>
-        </extensions>
+PHPUnit currently ships with 64 immutable event classes and corresponding event subscriber interfaces.
 
-        <!-- ... -->
-    </phpunit>
+Implementing a PHPUnit event subscriber
+---------------------------------------
 
-...
+An event subscriber is a class that implements an event subscriber interface.
+
+An event subscriber interface declares a single ``notify()`` method that accepts an instance of the corresponding event class.
+
+.. literalinclude:: examples/extending-phpunit/ExampleSubscriber.php
+   :caption: An ExampleSubscriber printing a message when PHPUnit emits the ExecutionFinished event
+   :language: php
+
+After registering an event subscriber with the extension facade, PHPUnit will notify the event subscriber when emitting an event of the corresponding event class.
+
+.. Note::
+
+  You can not create an event subscriber that implements more than one event subscriber interface at a time.
+
+  If you want to subscribe to more than one event, you need to implement at least one event subscriber for each event you are interested in.
+
+Implementing a PHPUnit event tracer
+-----------------------------------
+
+An event tracer is a class that implements the ``PHPUnit\Event\Tracer\Tracer`` interface.
+
+The tracer interface declares a single ``trace()`` method that accepts an event.
+
+.. literalinclude:: examples/extending-phpunit/ExampleTracer.php
+   :caption: An ExampleTracer receiving all events
+   :language: php
+
+After registering an event tracer with the extension facade, PHPUnit will notify the tracer of every event.
+
+.. Hint::
+
+  Are you unsure whether you should implement an event tracer or multiple event subscribers?
+
+  If you are interested in every event that PHPUnit emits during the execution of the CLI application, you probably want to implement and register an event tracer.
+
+  If you are interested in selected events that PHPUnit emits during the execution of the CLI application, you probably want to implement and register one or more event subscribers.
+
+Registering a PHPUnit extension
+-------------------------------
+
+You can register one or more PHPUnit extensions using the :ref:`extensions <appendixes.configuration.extensions>`, :ref:`bootstrap <appendixes.configuration.extensions.bootstrap>`, and :ref:`parameters <appendixes.configuration.extensions.extension.arguments>` elements of the :ref:`PHPUnit XML configuration file <appendixes.configuration>`.
+
+.. literalinclude:: examples/extending-phpunit/phpunit.xml
+   :caption: An XML configuration registering the ExampleExtension with parameters
+   :language: xml
 
 .. _extending-phpunit.event-system.event-system:
 
 PHPUnit's Event System
 ----------------------
 
```

Comment:
I am not sure about this section, perhaps it would make more sense to add an **Events** appendix listing all events?

It's also a bit difficult to find a suitable order for events (we briefly talked about it at IPC 2023 in Berlin). With an appendix, we could keep events sorted by name, what do you think?

Reply:
Sure, we can put them into the appendix. However, sorting them by name makes no sense to me. At the very least, we should sort them by "topic": test runner, test suite, test, etc.

In the [code](https://github.com/sebastianbergmann/phpunit/blob/main/src/Event/Emitter/Emitter.php) I have (tried to) sort(ed) them "chronologically". Don't know how useful that is.